// decaffeinate/dist/stages/main/patchers/BlockPatcher.mjs
import * as tslib_1 from "tslib";
import { SourceType } from 'coffee-lex';
import SharedBlockPatcher from '../../../patchers/SharedBlockPatcher';
import countVariableUsages from '../../../utils/countVariableUsages';
import notNull from '../../../utils/notNull';
import FunctionPatcher from './FunctionPatcher';
import ReturnPatcher from './ReturnPatcher';
var BlockPatcher = (function (_super) {
    tslib_1.__extends(BlockPatcher, _super);
    function BlockPatcher() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this._forPatcherDescendants = [];
        _this._explicitDeclarationsToAdd = [];
        return _this;
    }
    /**
     * Called by initialize within child ForPatchers, so this array will be
     * available in our initialize method.
     */
    BlockPatcher.prototype.markForPatcherDescendant = function (forPatcher) {
        this._forPatcherDescendants.push(forPatcher);
    };
    /**
     * In some cases, loops assign to variables, but will be turned into IIFEs,
     * moving the variable scope into the IIFE. This can cause incorrect behavior
     * if the variable is used outside the loop body, so we want to explicitly
     * declare the variable at the top of the block in that case.
     */
    BlockPatcher.prototype.initialize = function () {
        try {
            for (var _a = tslib_1.__values(this._forPatcherDescendants), _b = _a.next(); !_b.done; _b = _a.next()) {
                var forPatcher = _b.value;
                try {
                    for (var _c = tslib_1.__values(forPatcher.getIIFEAssignments()), _d = _c.next(); !_d.done; _d = _c.next()) {
                        var name = _d.value;
                        if (countVariableUsages(this.node, name) > countVariableUsages(forPatcher.node, name) &&
                            this._explicitDeclarationsToAdd.indexOf(name) === -1) {
                            this._explicitDeclarationsToAdd.push(name);
                        }
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (_d && !_d.done && (_e = _c.return)) _e.call(_c);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_b && !_b.done && (_f = _a.return)) _f.call(_a);
            }
            finally { if (e_2) throw e_2.error; }
        }
        var e_2, _f, e_1, _e;
    };
    BlockPatcher.prototype.canPatchAsExpression = function () {
        return this._explicitDeclarationsToAdd.length === 0 &&
            this.statements.every(function (statement) { return statement.canPatchAsExpression(); });
    };
    BlockPatcher.prototype.prefersToPatchAsExpression = function () {
        return this.statements.length === 0 ||
            (this.statements.length === 1 && this.statements[0].prefersToPatchAsExpression());
    };
    BlockPatcher.prototype.setExpression = function (force) {
        if (force === void 0) { force = false; }
        var willPatchAsExpression = _super.prototype.setExpression.call(this, force);
        if (willPatchAsExpression && this.prefersToPatchAsExpression()) {
            this.statements.forEach(function (statement) { return statement.setExpression(); });
            return true;
        }
        return false;
    };
    BlockPatcher.prototype.setImplicitlyReturns = function () {
        // A block can have no statements if it only had a block comment.
        if (this.statements.length > 0) {
            this.statements[this.statements.length - 1].setImplicitlyReturns();
        }
    };
    /**
     * Force the patcher to treat the block as inline (semicolon-separated
     * statements) or not (newline-separated statements).
     */
    BlockPatcher.prototype.setShouldPatchInline = function (shouldPatchInline) {
        this.shouldPatchInline = shouldPatchInline;
    };
    BlockPatcher.prototype.patchAsStatement = function (_a) {
        var _this = this;
        var _b = _a === void 0 ? {} : _a, _c = _b.leftBrace, leftBrace = _c === void 0 ? true : _c, _d = _b.rightBrace, rightBrace = _d === void 0 ? true : _d;
        if (leftBrace) {
            this.insert(this.innerStart, '{');
        }
        if (this._explicitDeclarationsToAdd.length > 0) {
            this.insertStatementsAtIndex(["var " + this._explicitDeclarationsToAdd.join(', ') + ";"], 0);
        }
        var constructor = null;
        this.statements.forEach(function (statement, i, statements) {
            if (i === statements.length - 1 && _this.parent instanceof FunctionPatcher) {
                if (statement instanceof ReturnPatcher && !statement.expression) {
                    _this.removeFinalEmptyReturn(statement);
                    return;
                }
            }
            // If we see a constructor (which only happens when this is a class
            // block), defer it until the end. Its patching may need other class
            // keys to already be patched so that it can generate method binding
            // statements within the constructor.
            // Check against the 'Constructor' node type instead of doing
            // `instanceof` to avoid a circular import issue.
            if (statement.node.type === 'Constructor') {
                if (constructor) {
                    throw _this.error('Unexpectedly found two constructors in the same block.');
                }
                constructor = statement;
            }
            else {
                _this.patchInnerStatement(statement);
            }
        });
        if (constructor) {
            this.patchInnerStatement(constructor);
        }
        if (rightBrace) {
            if (this.inline()) {
                this.insert(this.innerEnd, ' }');
            }
            else {
                this.appendLineAfter('}', -1);
            }
        }
    };
    BlockPatcher.prototype.patchInnerStatement = function (statement) {
        var hasImplicitReturn = (statement.implicitlyReturns() &&
            !statement.explicitlyReturns());
        if (statement.isSurroundedByParentheses() &&
            !statement.statementNeedsParens() &&
            !hasImplicitReturn) {
            this.remove(statement.outerStart, statement.innerStart);
            this.remove(statement.innerEnd, statement.outerEnd);
        }
        var implicitReturnPatcher = hasImplicitReturn ?
            this.implicitReturnPatcher() : null;
        if (implicitReturnPatcher) {
            implicitReturnPatcher.patchImplicitReturnStart(statement);
        }
        statement.patch();
        if (implicitReturnPatcher) {
            implicitReturnPatcher.patchImplicitReturnEnd(statement);
        }
        if (statement.statementNeedsSemicolon()) {
            this.insert(statement.outerEnd, ';');
        }
    };
    /**
     * Remove an unnecessary empty return at the end of a function. Ideally, we
     * want to remove the whole line, but that's only safe if the `return` is on a
     * line by itself. Otherwise, there might be bugs like code being pulled into
     * a comment on the previous line.
     */
    BlockPatcher.prototype.removeFinalEmptyReturn = function (statement) {
        var previousTokenIndex = statement.contentStartTokenIndex.previous();
        var nextTokenIndex = statement.contentEndTokenIndex.next();
        var previousToken = previousTokenIndex && this.sourceTokenAtIndex(previousTokenIndex);
        var nextToken = nextTokenIndex && this.sourceTokenAtIndex(nextTokenIndex);
        if (previousToken && previousToken.type === SourceType.NEWLINE &&
            (!nextToken || nextToken.type === SourceType.NEWLINE)) {
            this.remove(previousToken.start, statement.outerEnd);
        }
        else if (previousToken && previousToken.type === SourceType.SEMICOLON) {
            this.remove(previousToken.start, statement.outerEnd);
        }
        else {
            this.remove(statement.outerStart, statement.outerEnd);
        }
    };
    BlockPatcher.prototype.patchAsExpression = function (_a) {
        var _this = this;
        var _b = _a === void 0 ? {} : _a, _c = _b.leftBrace, leftBrace = _c === void 0 ? this.statements.length > 1 : _c, _d = _b.rightBrace, rightBrace = _d === void 0 ? this.statements.length > 1 : _d;
        if (leftBrace) {
            this.insert(this.innerStart, '(');
        }
        if (this.statements.length === 0) {
            this.insert(this.contentStart, 'undefined');
        }
        else {
            this.statements.forEach(function (statement, i, statements) {
                statement.setRequiresExpression();
                statement.patch();
                if (i !== statements.length - 1) {
                    var semicolonTokenIndex = _this.getSemicolonSourceTokenIndexBetween(statement, statements[i + 1]);
                    if (semicolonTokenIndex) {
                        var semicolonToken = notNull(_this.sourceTokenAtIndex(semicolonTokenIndex));
                        _this.overwrite(semicolonToken.start, semicolonToken.end, ',');
                    }
                    else {
                        _this.insert(statement.outerEnd, ',');
                    }
                }
            });
        }
        var lastToken = this.lastToken();
        if (lastToken.type === SourceType.SEMICOLON) {
            this.remove(lastToken.start, lastToken.end);
        }
        if (rightBrace) {
            this.insert(this.innerEnd, ')');
        }
    };
    /**
     * @private
     */
    BlockPatcher.prototype.getSemicolonSourceTokenIndexBetween = function (left, right) {
        return this.indexOfSourceTokenBetweenPatchersMatching(left, right, function (token) { return token.type === SourceType.SEMICOLON; });
    };
    /**
     * Blocks only exit via the last statement, so we check its code paths.
     */
    BlockPatcher.prototype.allCodePathsPresent = function () {
        return this.statements[this.statements.length - 1].allCodePathsPresent();
    };
    return BlockPatcher;
}(SharedBlockPatcher));
export default BlockPatcher;
