// decaffeinate/dist/stages/main/patchers/ChainedComparisonOpPatcher.mjs
import * as tslib_1 from "tslib";
import getCompareOperator from '../../../utils/getCompareOperator';
import isCompareOpNegationUnsafe from '../../../utils/isCompareOpNegationUnsafe';
import NodePatcher from './../../../patchers/NodePatcher';
/**
 * Handles constructs of the form `a < b < c < … < z`.
 */
var ChainedComparisonOpPatcher = (function (_super) {
    tslib_1.__extends(ChainedComparisonOpPatcher, _super);
    /**
     * `node` should have type `ChainedComparisonOp`.
     */
    function ChainedComparisonOpPatcher(patcherContext, operands) {
        var _this = _super.call(this, patcherContext) || this;
        _this.negated = false;
        _this.operands = operands;
        return _this;
    }
    ChainedComparisonOpPatcher.prototype.initialize = function () {
        try {
            for (var _a = tslib_1.__values(this.operands), _b = _a.next(); !_b.done; _b = _a.next()) {
                var operand = _b.value;
                operand.setRequiresExpression();
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
            }
            finally { if (e_1) throw e_1.error; }
        }
        var e_1, _c;
    };
    ChainedComparisonOpPatcher.prototype.patchAsExpression = function (_a) {
        var _b = (_a === void 0 ? {} : _a).needsParens, needsParens = _b === void 0 ? false : _b;
        var negateEntireExpression = this.shouldNegateEntireExpression();
        var addParens = negateEntireExpression ||
            (needsParens && !this.isSurroundedByParentheses());
        if (negateEntireExpression) {
            this.insert(this.contentStart, '!');
        }
        if (addParens) {
            this.insert(this.contentStart, '(');
        }
        var middle = this.getMiddleOperands();
        var negated = !negateEntireExpression && this.negated;
        var logicalOperator = negated ? '||' : '&&';
        try {
            for (var middle_1 = tslib_1.__values(middle), middle_1_1 = middle_1.next(); !middle_1_1.done; middle_1_1 = middle_1.next()) {
                var operand = middle_1_1.value;
                operand.setRequiresRepeatableExpression({ parens: true, ref: 'middle' });
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (middle_1_1 && !middle_1_1.done && (_c = middle_1.return)) _c.call(middle_1);
            }
            finally { if (e_2) throw e_2.error; }
        }
        try {
            for (var _d = tslib_1.__values(this.operands.entries()), _e = _d.next(); !_e.done; _e = _d.next()) {
                var _f = tslib_1.__read(_e.value, 2), i = _f[0], operand = _f[1];
                operand.patch();
                var operator = this.node.operators[i];
                if (operator) {
                    var replacement = getCompareOperator(operator.operator, negated);
                    if (operator.operator !== replacement) {
                        this.overwrite(operator.token.start, operator.token.end, replacement);
                    }
                }
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (_e && !_e.done && (_g = _d.return)) _g.call(_d);
            }
            finally { if (e_3) throw e_3.error; }
        }
        try {
            for (var middle_2 = tslib_1.__values(middle), middle_2_1 = middle_2.next(); !middle_2_1.done; middle_2_1 = middle_2.next()) {
                var operand = middle_2_1.value;
                // `a < b < c` → `a < b && b < c`
                //                     ^^^^^
                this.insert(operand.outerEnd, " " + logicalOperator + " " + operand.getRepeatCode());
            }
        }
        catch (e_4_1) { e_4 = { error: e_4_1 }; }
        finally {
            try {
                if (middle_2_1 && !middle_2_1.done && (_h = middle_2.return)) _h.call(middle_2);
            }
            finally { if (e_4) throw e_4.error; }
        }
        if (addParens) {
            this.insert(this.contentEnd, ')');
        }
        var e_2, _c, e_3, _g, e_4, _h;
    };
    /**
     * If any negation is unsafe, just wrap the whole thing in parens with a !
     * operator. That's easier and arguably nicer-looking than trying to
     * intelligently negate the subexpressions accounting for unsafe negations.
     */
    ChainedComparisonOpPatcher.prototype.shouldNegateEntireExpression = function () {
        return this.negated &&
            this.node.operators.some(function (operator) { return isCompareOpNegationUnsafe(operator.operator); }) &&
            !this.options.looseComparisonNegation;
    };
    /**
     * @private
     */
    ChainedComparisonOpPatcher.prototype.getMiddleOperands = function () {
        return this.operands.slice(1, -1);
    };
    ChainedComparisonOpPatcher.prototype.negate = function () {
        this.negated = !this.negated;
    };
    /**
     * Forward the request to the first operand.
     */
    ChainedComparisonOpPatcher.prototype.statementNeedsParens = function () {
        return this.operands[0].statementShouldAddParens();
    };
    return ChainedComparisonOpPatcher;
}(NodePatcher));
export default ChainedComparisonOpPatcher;
