// decaffeinate/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var decaffeinate_coffeescript_1 = require("decaffeinate-coffeescript");
var index_1 = require("./stages/add-variable-declarations/index");
var index_2 = require("./stages/esnext/index");
var index_3 = require("./stages/literate/index");
var index_4 = require("./stages/main/index");
var index_5 = require("./stages/normalize/index");
var index_6 = require("./stages/semicolons/index");
var suggestions_1 = require("./suggestions");
var convertNewlines_1 = require("./utils/convertNewlines");
var DecaffeinateContext_1 = require("./utils/DecaffeinateContext");
var detectNewlineStr_1 = require("./utils/detectNewlineStr");
var formatCoffeeLexTokens_1 = require("./utils/formatCoffeeLexTokens");
var formatCoffeeScriptAst_1 = require("./utils/formatCoffeeScriptAst");
var formatCoffeeScriptLexerTokens_1 = require("./utils/formatCoffeeScriptLexerTokens");
var formatDecaffeinateParserAst_1 = require("./utils/formatDecaffeinateParserAst");
var PatchError_1 = require("./utils/PatchError");
exports.PatchError = PatchError_1.default;
var removeUnicodeBOMIfNecessary_1 = require("./utils/removeUnicodeBOMIfNecessary");
var resolveToPatchError_1 = require("./utils/resolveToPatchError");
var cli_1 = require("./cli");
exports.run = cli_1.default;
var options_1 = require("./options");
var notNull_1 = require("./utils/notNull");
/**
 * Convert CoffeeScript source code into modern JavaScript preserving comments
 * and formatting.
 */
function convert(source, options) {
    if (options === void 0) { options = {}; }
    source = removeUnicodeBOMIfNecessary_1.default(source);
    options = Object.assign({}, options_1.DEFAULT_OPTIONS, options);
    var originalNewlineStr = detectNewlineStr_1.default(source);
    source = convertNewlines_1.default(source, '\n');
    var literate = options.literate ||
        notNull_1.default(options.filename).endsWith('.litcoffee') ||
        notNull_1.default(options.filename).endsWith('.coffee.md');
    var stages = tslib_1.__spread((literate ? [index_3.default] : []), [
        index_5.default,
        index_4.default,
        index_1.default,
        index_6.default,
        index_2.default
    ]);
    var runToStage = options.runToStage;
    if (runToStage !== null && runToStage !== undefined) {
        var stageIndex = stages.findIndex(function (stage) { return stage.name === runToStage; });
        if (stageIndex !== -1) {
            stages = stages.slice(0, stageIndex + 1);
        }
        else {
            return convertCustomStage(source, runToStage);
        }
    }
    var result = runStages(source, options, stages);
    if (!options.disableSuggestionComment) {
        result.code = suggestions_1.prependSuggestionComment(result.code, result.suggestions);
    }
    result.code = convertNewlines_1.default(result.code, originalNewlineStr);
    return {
        code: result.code,
    };
}
exports.convert = convert;
function modernizeJS(source, options) {
    if (options === void 0) { options = {}; }
    source = removeUnicodeBOMIfNecessary_1.default(source);
    options = Object.assign({}, options_1.DEFAULT_OPTIONS, options);
    var originalNewlineStr = detectNewlineStr_1.default(source);
    source = convertNewlines_1.default(source, '\n');
    var stages = [
        index_2.default
    ];
    var result = runStages(source, options, stages);
    result.code = convertNewlines_1.default(result.code, originalNewlineStr);
    return {
        code: result.code,
    };
}
exports.modernizeJS = modernizeJS;
function runStages(initialContent, options, stages) {
    var content = initialContent;
    var suggestions = [];
    stages.forEach(function (stage) {
        var _a = runStage(stage, content, options), code = _a.code, stageSuggestions = _a.suggestions;
        content = code;
        suggestions.push.apply(suggestions, tslib_1.__spread(stageSuggestions));
    });
    return { code: content, suggestions: suggestions_1.mergeSuggestions(suggestions) };
}
function runStage(stage, content, options) {
    try {
        return stage.run(content, options);
    }
    catch (err) {
        var patchError = resolveToPatchError_1.default(err, content, stage.name);
        if (patchError !== null) {
            throw patchError;
        }
        throw err;
    }
}
function convertCustomStage(source, stageName) {
    var context = DecaffeinateContext_1.default.create(source);
    if (stageName === 'coffeescript-lexer') {
        return {
            code: formatCoffeeScriptLexerTokens_1.default(decaffeinate_coffeescript_1.tokens(source), context),
        };
    }
    else if (stageName === 'coffeescript-parser') {
        return {
            code: formatCoffeeScriptAst_1.default(context),
        };
    }
    else if (stageName === 'coffee-lex') {
        return {
            code: formatCoffeeLexTokens_1.default(context),
        };
    }
    else if (stageName === 'decaffeinate-parser') {
        return {
            code: formatDecaffeinateParserAst_1.default(context),
        };
    }
    else {
        throw new Error("Unrecognized stage name: " + stageName);
    }
}
