// decaffeinate/dist/index.mjs
import * as tslib_1 from "tslib";
import { tokens } from 'decaffeinate-coffeescript';
import AddVariableDeclarationsStage from './stages/add-variable-declarations/index';
import EsnextStage from './stages/esnext/index';
import LiterateStage from './stages/literate/index';
import MainStage from './stages/main/index';
import NormalizeStage from './stages/normalize/index';
import SemicolonsStage from './stages/semicolons/index';
import { mergeSuggestions, prependSuggestionComment } from './suggestions';
import convertNewlines from './utils/convertNewlines';
import DecaffeinateContext from './utils/DecaffeinateContext';
import detectNewlineStr from './utils/detectNewlineStr';
import formatCoffeeLexAst from './utils/formatCoffeeLexTokens';
import formatCoffeeScriptAst from './utils/formatCoffeeScriptAst';
import formatCoffeeScriptLexerTokens from './utils/formatCoffeeScriptLexerTokens';
import formatDecaffeinateParserAst from './utils/formatDecaffeinateParserAst';
import PatchError from './utils/PatchError';
import removeUnicodeBOMIfNecessary from './utils/removeUnicodeBOMIfNecessary';
import resolveToPatchError from './utils/resolveToPatchError';
export { default as run } from './cli';
import { DEFAULT_OPTIONS } from './options';
import notNull from './utils/notNull';
export { PatchError };
/**
 * Convert CoffeeScript source code into modern JavaScript preserving comments
 * and formatting.
 */
export function convert(source, options) {
    if (options === void 0) { options = {}; }
    source = removeUnicodeBOMIfNecessary(source);
    options = Object.assign({}, DEFAULT_OPTIONS, options);
    var originalNewlineStr = detectNewlineStr(source);
    source = convertNewlines(source, '\n');
    var literate = options.literate ||
        notNull(options.filename).endsWith('.litcoffee') ||
        notNull(options.filename).endsWith('.coffee.md');
    var stages = tslib_1.__spread((literate ? [LiterateStage] : []), [
        NormalizeStage,
        MainStage,
        AddVariableDeclarationsStage,
        SemicolonsStage,
        EsnextStage
    ]);
    var runToStage = options.runToStage;
    if (runToStage !== null && runToStage !== undefined) {
        var stageIndex = stages.findIndex(function (stage) { return stage.name === runToStage; });
        if (stageIndex !== -1) {
            stages = stages.slice(0, stageIndex + 1);
        }
        else {
            return convertCustomStage(source, runToStage);
        }
    }
    var result = runStages(source, options, stages);
    if (!options.disableSuggestionComment) {
        result.code = prependSuggestionComment(result.code, result.suggestions);
    }
    result.code = convertNewlines(result.code, originalNewlineStr);
    return {
        code: result.code,
    };
}
export function modernizeJS(source, options) {
    if (options === void 0) { options = {}; }
    source = removeUnicodeBOMIfNecessary(source);
    options = Object.assign({}, DEFAULT_OPTIONS, options);
    var originalNewlineStr = detectNewlineStr(source);
    source = convertNewlines(source, '\n');
    var stages = [
        EsnextStage
    ];
    var result = runStages(source, options, stages);
    result.code = convertNewlines(result.code, originalNewlineStr);
    return {
        code: result.code,
    };
}
function runStages(initialContent, options, stages) {
    var content = initialContent;
    var suggestions = [];
    stages.forEach(function (stage) {
        var _a = runStage(stage, content, options), code = _a.code, stageSuggestions = _a.suggestions;
        content = code;
        suggestions.push.apply(suggestions, tslib_1.__spread(stageSuggestions));
    });
    return { code: content, suggestions: mergeSuggestions(suggestions) };
}
function runStage(stage, content, options) {
    try {
        return stage.run(content, options);
    }
    catch (err) {
        var patchError = resolveToPatchError(err, content, stage.name);
        if (patchError !== null) {
            throw patchError;
        }
        throw err;
    }
}
function convertCustomStage(source, stageName) {
    var context = DecaffeinateContext.create(source);
    if (stageName === 'coffeescript-lexer') {
        return {
            code: formatCoffeeScriptLexerTokens(tokens(source), context),
        };
    }
    else if (stageName === 'coffeescript-parser') {
        return {
            code: formatCoffeeScriptAst(context),
        };
    }
    else if (stageName === 'coffee-lex') {
        return {
            code: formatCoffeeLexAst(context),
        };
    }
    else if (stageName === 'decaffeinate-parser') {
        return {
            code: formatDecaffeinateParserAst(context),
        };
    }
    else {
        throw new Error("Unrecognized stage name: " + stageName);
    }
}
