// coffee-lex/dist/SourceType.js
"use strict";
exports.__esModule = true;
/**
 * Represents a particular type of CoffeeScript code.
 */
var SourceType;
(function (SourceType) {
    SourceType["AT"] = "AT";
    SourceType["BOOL"] = "BOOL";
    SourceType["BREAK"] = "BREAK";
    SourceType["CATCH"] = "CATCH";
    SourceType["CALL_END"] = "CALL_END";
    SourceType["CALL_START"] = "CALL_START";
    SourceType["CLASS"] = "CLASS";
    SourceType["COLON"] = "COLON";
    SourceType["COMMA"] = "COMMA";
    SourceType["COMMENT"] = "COMMENT";
    SourceType["CONTINUATION"] = "CONTINUATION";
    SourceType["CONTINUE"] = "CONTINUE";
    SourceType["DECREMENT"] = "DECREMENT";
    SourceType["DELETE"] = "DELETE";
    SourceType["DO"] = "DO";
    SourceType["DOT"] = "DOT";
    SourceType["DSTRING_START"] = "DSTRING_START";
    SourceType["DSTRING_END"] = "DSTRING_END";
    SourceType["ELSE"] = "ELSE";
    SourceType["EOF"] = "EOF";
    SourceType["EXISTENCE"] = "EXISTENCE";
    SourceType["EXTENDS"] = "EXTENDS";
    SourceType["FINALLY"] = "FINALLY";
    SourceType["FOR"] = "FOR";
    SourceType["FUNCTION"] = "FUNCTION";
    SourceType["HERECOMMENT"] = "HERECOMMENT";
    SourceType["HEREGEXP_START"] = "HEREGEXP_START";
    SourceType["HEREGEXP_END"] = "HEREGEXP_END";
    SourceType["IF"] = "IF";
    SourceType["INCREMENT"] = "INCREMENT";
    SourceType["INTERPOLATION_START"] = "INTERPOLATION_START";
    SourceType["INTERPOLATION_END"] = "INTERPOLATION_END";
    SourceType["JS"] = "JS";
    SourceType["LBRACE"] = "LBRACE";
    SourceType["LBRACKET"] = "LBRACKET";
    SourceType["LOOP"] = "LOOP";
    SourceType["LPAREN"] = "LPAREN";
    SourceType["NEW"] = "NEW";
    SourceType["NEWLINE"] = "NEWLINE";
    SourceType["NORMAL"] = "NORMAL";
    SourceType["NULL"] = "NULL";
    SourceType["NUMBER"] = "NUMBER";
    SourceType["OPERATOR"] = "OPERATOR";
    SourceType["OWN"] = "OWN";
    SourceType["PROTO"] = "PROTO";
    SourceType["RANGE"] = "RANGE";
    SourceType["REGEXP"] = "REGEXP";
    SourceType["RBRACE"] = "RBRACE";
    SourceType["RBRACKET"] = "RBRACKET";
    SourceType["RELATION"] = "RELATION";
    SourceType["RETURN"] = "RETURN";
    SourceType["RPAREN"] = "RPAREN";
    SourceType["SEMICOLON"] = "SEMICOLON";
    SourceType["SPACE"] = "SPACE";
    SourceType["SUPER"] = "SUPER";
    SourceType["SWITCH"] = "SWITCH";
    SourceType["SSTRING_START"] = "SSTRING_START";
    SourceType["SSTRING_END"] = "SSTRING_END";
    SourceType["STRING_CONTENT"] = "STRING_CONTENT";
    SourceType["STRING_LINE_SEPARATOR"] = "STRING_LINE_SEPARATOR";
    SourceType["STRING_PADDING"] = "STRING_PADDING";
    SourceType["TDSTRING_START"] = "TDSTRING_START";
    SourceType["TDSTRING_END"] = "TDSTRING_END";
    SourceType["THEN"] = "THEN";
    SourceType["THIS"] = "THIS";
    SourceType["THROW"] = "THROW";
    SourceType["TRY"] = "TRY";
    SourceType["TSSTRING_START"] = "TSSTRING_START";
    SourceType["TSSTRING_END"] = "TSSTRING_END";
    SourceType["UNDEFINED"] = "UNDEFINED";
    SourceType["UNKNOWN"] = "UNKNOWN";
    SourceType["WHEN"] = "WHEN";
    SourceType["WHILE"] = "WHILE";
    SourceType["IDENTIFIER"] = "IDENTIFIER";
    SourceType["YIELD"] = "YIELD";
    SourceType["YIELDFROM"] = "YIELDFROM";
})(SourceType || (SourceType = {}));
exports["default"] = SourceType;
