// decaffeinate/dist/stages/semicolons/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var buildConfig = require('ast-processor-babylon-config');
var asi = require('automatic-semicolon-insertion');
var babylon_1 = require("babylon");
var magic_string_1 = require("magic-string");
var debug_1 = require("../../utils/debug");
var BABYLON_PLUGINS = [
    'flow',
    'jsx',
    'asyncGenerators',
    'classConstructorCall',
    'classProperties',
    'decorators',
    'doExpressions',
    'exportExtensions',
    'functionBind',
    'functionSent',
    'objectRestSpread',
    'optionalChaining',
];
var SemicolonsStage = (function () {
    function SemicolonsStage() {
    }
    SemicolonsStage.run = function (content) {
        var log = debug_1.logger(this.name);
        log(content);
        var editor = new magic_string_1.default(content);
        var ast = babylon_1.parse(content, {
            sourceType: 'module',
            plugins: BABYLON_PLUGINS,
            allowReturnOutsideFunction: true,
            tokens: true,
        }); // tslint:disable-line no-any
        var config = buildConfig(content, ast);
        asi(config);
        config.insertions.forEach(function (_a) {
            var index = _a.index, content = _a.content;
            return editor.appendLeft(index, content);
        });
        config.removals.forEach(function (_a) {
            var start = _a.start, end = _a.end;
            return editor.remove(start, end);
        });
        return {
            code: editor.toString(),
            suggestions: [],
        };
    };
    return SemicolonsStage;
}());
exports.default = SemicolonsStage;
